src/detect-mqtt-connack-sessionpresent.c

```diff
@@ -0,0 +1,313 @@
+ * \brief Function to match session_present flag of an MQTT CONNACK message
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectMQTTConnackSessionPresentData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTConnackSessionPresentMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const bool *de = (const bool *)ctx;
+    bool value = false;
+
+    if (!de)
+        return 0;
+
+    if (rs_mqtt_tx_get_connack_sessionpresent(txv, &value) ==0 ) {
+        return 0;
+    }
+    if (value != *de) {
+        return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * \internal
+ * \brief This function is used to parse options passed via mqtt.connack.session_present: keyword
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTConnackSessionPresentData on success
+ * \retval NULL on failure
+ */
+static bool *DetectMQTTConnackSessionPresentParse(const char *rawstr)
+{
+    bool *de = NULL;
+#define MAX_SUBSTRINGS 30
+    de = SCMalloc(sizeof(bool));
+    if (unlikely(de == NULL))
+        return NULL;
+    *de = false;
+
+    if (strcmp(rawstr, "yes") == 0) {
+        *de = true;
+    } else if (strcmp(rawstr, "true") == 0) {
+        *de = true;
+    } else if (strcmp(rawstr, "no") == 0) {
+        *de = false;
+    } else if (strcmp(rawstr, "false") == 0) {
+        *de = false;
+    } else {
+        SCLogError(SC_ERR_UNKNOWN_VALUE, "invalid session_present flag definition: %s", rawstr);
+        goto error;
+    }
+
+    return de;
+
+error:
+    SCFree(de);
+    return NULL;
+}
+
+/**
+ * \internal
+ * \brief this function is used to add the parsed type query into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param rawstr pointer to the user provided options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectMQTTConnackSessionPresentSetup (DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+{
+    bool *de = NULL;
+    SigMatch *sm = NULL;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_MQTT) < 0)
+        return -1;
+
+    de = DetectMQTTConnackSessionPresentParse(rawstr);
+    if (de == NULL)
+        goto error;
+
+    sm = SigMatchAlloc();
+    if (sm == NULL)
+        goto error;
+
+    sm->type = DETECT_AL_MQTT_CONNACK_SESSION_PRESENT;
+    sm->ctx = (SigMatchCtx *)de;
+
+    SigMatchAppendSMToList(s, sm, mqtt_connack_session_present_id);
+
+    return 0;
+
+error:
+    SCFree(de);
+    SCFree(sm);
```

Comment:
Suggest guarding each free with `if (<var-name> != NULL)`

Reply:
Same comment for other detect modules.

Reply:
Interesting. Is there anything to say against just having `SCFree()` do nothing if the passed pointer is `NULL`?

Reply:
Done, BTW :)

Reply:
I think it is a coding pattern (used in other projects as well) to avoid double frees and leaks

By the way, `SCFree` does nothing if the passed pointer is NULL